AVAS π Orbital Selection (#261)
evangelistalab/forte: docs/source/methods/avas.rst

^^^^^^^^^^^^^

> In this example, we perform the AVAS procedure on formaldehyde

Reviewer: ```suggestion
In this example, we use AVAS to find an active space for formaldehyde that spans the :math:`2p_x` orbitals of the C and O atoms,
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For the above example, the molecule is placed on the yz plane such that the directions of p orbitals
are aligned with the molecule xyz frame.
However, sometimes it is hard to make these two xyz frames aligned.
For example, the molecule contains multiple π subsystems or the plane is not perfect.
In such cases, if only pi orbitals are desired, we can linearly combine the px, py, and pz orbitals
to a single 'pz' orbital that perpendicular to the π subsystem plane.

We use the keyword :code:`SUBSPACE_PI_PLANES` to define all the π subsystems planes,
where each plane is given by a list of atoms' expressions.
Some valid expressions would be: ::

  - [['C', 'H', 'O']]  # only one plane consisting all C, H, and O atoms of the molecule.
  - [['C1-6'], ['N1-2', 'C9-11']]  # plane 1 with the first six C atoms of the molecule,
                                   # plane 2 with C9, C10, C11, N1 and N2 atoms.
  - [['C1-4'], ['C1-2', 'C5-6']]:  # plane 1 with the first four C atoms of the molecule,
                                   # plane 2 with C1, C2, C5, and C6 atoms.
                                   # Two planes share C1 and C2!

> 

Author: @rwashi690  I think you miss the point. Let me rewrite the manual.

This part is just to illustrate the syntax of that keyword.

What I want to achieve is very simple. So suppose using the provided formaldehyde geometry, which is not lying in any of the xy, zx, or yz plane. Now I want AVAS select the pi orbitals of the molecule, which consist of the C and O 2p orbitals that are perpendicular to the plane. Now how to choose these two p orbitals? They are not aligned to x, y, nor z direction. You could try to use ["C(2px)", "O(2px)"] and see what AVAS yields.

There are two ways for this particular case. We could try to rotate the formaldehyde molecule such that the molecular plane is located at xy, xz, or yz plane, and the 2p orbitals are simply aligned with the other direction. For example, if the molecule is at yz plane, the subspace orbitals would be ["C(2px)", "O(2px)"].

The other approach is not to move the molecule but to project all three 2p orbitals onto the plane normal. Suppose the plane normal is (nx, ny, nz). The 2p orbital perpendicular to the plane is simply nx * 2px + ny * 2py + nz * 2pz. This is what I mean px, py, pz orbitals goes to 'pz' orbital that is perpendicular to the plane. Note the quotation mark.

Which approach is more general? I must say it is the second because there might be multiple pi systems in a molecule.

Take the formaldehyde example above.
We now reorient it such that the plane normal points at (1.0, 1.0, 1.0).
The same AVAS orbitals can be obtained using the following input: ::

  molecule H2CO{
  0 1
  C        0.346146295209737    0.126698337466632   -0.472844632676369
  O       -0.346053592352928   -0.126664405871036    0.472717998223964
  H        1.227335215970831   -0.489581944167415   -0.737753271803415
  H        0.143281782803090    0.991262584630455   -1.134544367433545
  noreorient
  }

  set forte {

>     subspace           ["C(2p)", "O(2p)"]  # must include all p orbitals!

Author: See if my previous comment make sense.

>     subspace_pi_planes [["C", "O", "H"]]   # only one plane, defined by all C, O and H atoms

Author: The definition of plane has nothing to do orbital orbitals, but only atoms.

      set forte {
        minao_basis        double-shell
        subspace           ["C(2p)","C(3p)","O(2p)","O(3p)"]
        subspace_pi_planes [["C","O","H"]]
        avas               true
        avas_diagonalize   true
        avas_cutoff        0.5
      }

    You need to prepare your own MINAO basis (here I prepare a basis called "double-shell.gbs").
    This MINAO basis should include at least the 2p and 3p orbitals of C and O,
    which can be grabbed from the cc-pVTZ or ANO-RCC-VTZP basis sets.

For a more realistic example, consider the following iron porphyrin related molecule:

.. image:: images/FeP.png
    :width: 600
    :align: center
    :alt: An iron porphyrin complex.

This molecule contains two π systems, namely, porphyrin and imidazole.
Also, the porphyrin is not a perfect plane anymore.
The following input snippet selects all 3d orbitals of Fe, 3p orbitals of S,
and all 'pz' orbitals of porphyrin and imidazole rings. ::

  set forte {
    avas                true
    avas_diagonalize    true
    avas_cutoff         0.5
    minao_basis         cc-pvtz-minao
    subspace            ["Fe(3d)","C6-25(2p)","N(2p)","S(3p)","C1-3(2p)"]

>     subspace_pi_planes  [["Fe","C6-25","N3-6"], ["N1-2","C1-3"]]

Author: It is not possible to make such claim. Neither of them locate on xz or yz plane.

>     subspace_pi_planes  [["Fe","C6-25","N3-6"], ["N1-2","C1-3"]]

Reviewer: Still, I recommend to add a line to say that ["Fe","C6-25","N3-6"] defines the porphyrin plane, while ["N1-2","C1-3"] defines the imidazole ring. Otherwise the reader has to do the mental work of figuring out what these lists mean.

>     subspace_pi_planes  [["Fe","C6-25","N3-6"], ["N1-2","C1-3"]]

Author: Text added afterwards.

  }

The AVAS output selects exactly 37 orbitals we wanted. ::

  Sum of eigenvalues: 36.98863914

>   AVAS covers 96.59% of the subspace.

Reviewer 2: Does this come from the overlap matrix? If so SVD would be good to compute and get the average of the singular values

Author: This is the percentage of AVAS selected orbitals over the total subspace orbitals. The SVD values are actually printed afterwards.

For more examples, see :code:`avas-1` to :code:`avas-6` in the :code:`tests/methods` folder.
In particular, :code:`avas-6` is a practical example on ferrocene.

Defining the molecular plane for π orbitals
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Molecular systems with conjugated π bonds generally process planar geometries.

Reviewer: ```suggestion
Molecular systems with conjugated π bonds are often arranged into planar geometries.
```

> For such systems, the active orbitals are usually chosen as the π orbitals perpendicular to the plane.

Reviewer: ```suggestion
For such systems, it often desirable to select an active space that includes π orbitals perpendicular to the plane.
```

> Each π orbital consists of atomic p orbitals, which are also perpendicular to the plane.

Reviewer: ```suggestion
Each π orbital is a linear combination of atomic p orbitals, which are also perpendicular to the plane.
However, unlike the case of formaldehyde, where it easy to select the appropriate π and π* orbitals, in the more general case a π orbital is a linear combination of 2px, 2py, and 2px orbitals.

Fore example, in the following formaldehyde geometry, the atoms do not lie on the xy, xz, or yz plane ::
```

> Take the above formaldehyde as an example.
> The π and π* orbitals are built from the 2p orbitals of C and O that are perpendicular to the molecular plane.
> We are very lucky that the formaldehyde molecule is in yz plane and
> the subspace 2p orbitals can be easily chosen as the 2px orbitals of C and O atoms.

Reviewer: ```suggestion
```

> Take the above formaldehyde as an example.
> The π and π* orbitals are built from the 2p orbitals of C and O that are perpendicular to the molecular plane.
> We are very lucky that the formaldehyde molecule is in yz plane and
> the subspace 2p orbitals can be easily chosen as the 2px orbitals of C and O atoms.
> 
> In reality, such perfect alignment does not happen often.
> Consider the below formaldehyde geometry ::

Reviewer: ```suggestion
```

  H        1.227335215970831   -0.489581944167415   -0.737753271803415
  H        0.143281782803090    0.991262584630455   -1.134544367433545
  noreorient  # ask Psi4 not to reorient the xyz
  }

Now the plane normal points to (1.0, 1.0, 1.0).
Unfortunately Psi4 does not recognize the symmetry anymore unless we let it reorient the coordinate.
However, you should confirm that the bond lengths and angles are identical to those in the previous geometry.
If we use the same input for AVAS as above, here is what we get ::

  ==> Atomic Valence MOs (Active Marked by *) <==

    ===============================
     Irrep    MO  Occ.  <phi|P|phi>
    -------------------------------
    *   A      0    2      0.905072
    *   A      1    2      0.621379
    *   A      8    0      0.374453
    *   A      9    0      0.087267
    ===============================

Pay attention to the overlap metric.
The projector becomes less effective and it is unclear which is the π orbital.
The π* orbital also has a small overlap (0.374), much smaller than that from the previous aligned geometry (0.993).
The reason for this is simply the subspace orbitals (i.e., 2px of C and O)

> are not perpendicular to the molecular plane.

Reviewer: I think this example does not necessarily help. Sure, it explains what happens when you do the wrong thing and invoke AVAS on a nonplanar geometry. But this is something that I would explain pro-actively, in the sense that if we want the user to avoid an error, we should provide a list of few important things to do before running AVAS. I would jump right directly to the next section and directly say "The approach described in the previous section is not flexible enough to treat general π systems like molecules containing multiple π systems or π systems that are not aligned with a specific molecular axis."

There are two ways to fix this problem.
One is to reorient the molecule such that the molecular plane lying in yz plane.
However, this approach is not flexible enough to treat multiple π systems in a molecule.
The other approach is to use all px, py, pz orbitals as basis,
using which the p orbital perpendicular to the plane can be defined.
To do this, we need to specify two keywords: :code:`SUBSPACE` and :code:`SUBSPACE_PI_PLANES`.

> The latter is used to define the π plane, while we need all p orbitals for the former

Reviewer: ```suggestion
The latter is used to define the π plane, while we need all 2p orbitals for the former
```

Reviewer: I suggest adding some important details here.

  - ["C1-3"]           # carbon atoms #1, #2, #3
  - ["C(2p)"]          # the 2p subset of all carbon atoms
  - ["C(1s)","C(2s)"]  # the 1s/2s subsets of all carbon atoms
  - ["C1-3(2s)"]       # the 2s subsets of carbon atoms #1, #2, #3
  - ["Ce(4fzx2-zy2)"]  # the 4f zxx-zyy orbital of all Ce atoms

.. note::
    It is very important to include a complete set of p orbitals in :code:`SUBSPACE` if π planes are defined.
    Otherwise, the code will follow the directions given by :code:`SUBSPACE`.

.. tip::
    The code is flexible enough to treat double active spaces (e.g., double π or double d-shell).
    For example, the double-π active space of formaldehyde can be obtained via ::

      set forte {
        minao_basis        double-shell
        subspace           ["C(2p)","C(3p)","O(2p)","O(3p)"]
        subspace_pi_planes [["C","O","H"]]
        avas               true
        avas_diagonalize   true
        avas_cutoff        0.5
      }

    Here I prepare a basis called "double-shell.gbs", which includes the 2p and 3p orbitals of C and O atoms.
    You can also prepare your own MINAO basis by truncating the the cc-pVTZ or ANO-RCC-VTZP basis sets.

> For a more realistic example, consider the following iron porphyrin related molecule:

Reviewer: I think this should be a new section specifically for molecules with two or more pi planes. You also need to explain (perhaps at the end, how AVAS uses this information). There are a lot of technical details at the beginning of the text, but here we do not explain what the multiple planes do or why they are necessary.

Author: I added a new section. Hope the change matches your standard.

.. image:: images/FeP.png
    :width: 600
    :align: center
    :alt: An iron porphyrin complex.

This molecule contains two π systems, namely, porphyrin and imidazole.
Also, the porphyrin is not a perfect plane anymore.

> The following snippet selects all 3d orbitals of Fe, 3p orbitals of S,

Reviewer: ```suggestion
The following snippet selects all the 3d orbitals of Fe, the 3p orbitals of S,
```

and all p orbitals that are perpendicular of porphyrin and imidazole rings. ::

  set forte {
    avas                true
    avas_diagonalize    true
    avas_cutoff         0.5
    minao_basis         cc-pvtz-minao
    subspace            ["Fe(3d)","C6-25(2p)","N(2p)","S(3p)","C1-3(2p)"]
    subspace_pi_planes  [["Fe","C6-25","N3-6"], ["N1-2","C1-3"]]
  }

The AVAS output selects exactly 37 orbitals we wanted. ::

  ==> AVAS MOs Information <==

    ---------------------
                        A
    ---------------------
    DOCC INACTIVE     106
    DOCC ACTIVE        22
    SOCC ACTIVE         0
    UOCC ACTIVE        15
    UOCC INACTIVE     462
    ---------------------
    RESTRICTED_DOCC   106
    ACTIVE             37
    RESTRICTED_UOCC   462
    ---------------------

>   ==> Atomic Valence MOs (Active Marked by *) <==

Reviewer: Could we condense this table with lines that say "..."? Would the point we want to make be sacrificed if we shrink it?
